### ISSUE TITLE
Json: map non-primitive collections to JSON objects

### ISSUE THREAD

**Issue by maumar** [dotnet/efcore#28861]
Currently collections are mapped to JSON arrays, we could map them to JSON objects instead.

**Comment by roji** [comment]
@maumar can you provide some context on what this means?

**Comment by AndriySvyryd** [comment]
This would allow to map collections of owned types as objects where one of the properties on the owned type is mapped to the object key and the value contains another JSON object containing the rest of the properties.
In the client side it would still be a regular collection, as opposed to #21262

**Comment by roji** [comment]
Am noting that while this is technically possible, I'm not aware of anyone actual doing this (or requesting it). Significantly, AFAIK neither System.Text.Json nor Newtonsoft.Json support this mode (not out of the box in any case), which should be a pretty strong signal. If nobody's doing this with JSON outside of the database, I'm not sure why it would be needed in the database specifically.

Also, such a mapping cannot be ordered, since JSON objects are generally unordered, and binary representations of JSON (e.g. jsonb in PostgreSQL, but also in other databases) very explicitly do not preserve ordering. Even where they do, there is generally no operation for accessing the Nth key of an object, based on the object layout. We could, of course, support this as an unordered collection (similar to foreign-key-based relationships in non-JSON relational databases), but then the natural way to represent this would be via a Dictionary, no?

To summarize, I think it's an interesting theoretical thought, but I'm not sure what the purpose of such a mapping would be, and why users would want it.

**Comment by AndriySvyryd** [comment]
> Also, such a mapping cannot be ordered, since JSON objects are generally unordered, and binary representations of JSON (e.g. jsonb in PostgreSQL, but also in other databases) very explicitly do not preserve ordering. Even where they do, there is generally no operation for accessing the Nth key of an object, based on the object layout. We could, of course, support this as an unordered collection (similar to foreign-key-based relationships in non-JSON relational databases), but then the natural way to represent this would be via a Dictionary, no?

Precisely, this makes it the perfect mapping for a HashSet or Dictionary of entity types.

> To summarize, I think it's an interesting theoretical thought, but I'm not sure what the purpose of such a mapping would be, and why users would want it.

To enforce key uniqueness in the database with fast lookup by key value

**Comment by bachratyg** [comment]
> To enforce key uniqueness in the database with fast lookup by key value

FWIW the JSON spec does allow duplicate keys on an object but it's highly discouraged.

**Comment by roji** [comment]
> Precisely, this makes it the perfect mapping for a HashSet or Dictionary of entity types.

I see - the OP reads very much as if this is about allowing mapping e.g. .NET Lists/arrays to JSON objects ("Currently collections are mapped to JSON arrays.").

> To enforce key uniqueness in the database with fast lookup by key value

That may or may not work, based on the database implementation (@bachratyg's note goes in this direction). For example, on SQL Server with nvarchar mapping, duplicate keys would be persisted just fine. In fact, even in the new SQL Server 2025 JSON type, any duplicates are simply silently discarded:

```sql
SELECT CAST('{ "a" : 8, "a" : 9 }' AS json);
-- returns {"a":8}
```

Even where this works, it would enforce uniquness only within its collection, so duplicates would still exist across different JSON documents (and attempting to load both documents would fail in change tracking because of the duplicate key). Note that actual JSON uniqueness constraints in document databases (Mongo, Cosmos) are across all documents.

In any case, I do agree that mapping a Dictionary to a JSON object indeed seems very natural - regardless of key uniqueness or not (it's just a natural mapping); this is also of course supported by S.T.Json. I think that definitely makes sense to support.

However, mapping a HashSet makes much less sense to me. I see the idea in theory, but I think that in practice the uniqueness constraint doesn't actually work (as above), and IMHO we should not be in the business of arbitrary mapping transformations like this one; FWIW STJ does not support this out-of-the-box (in the spirit of using it as our oracle).

**Comment by AndriySvyryd** [comment]
The dictionary would need to have the key calculated based on the key properties of the related entity to allow fix-up to work. With this in mind, Dictionary and HashSet of entities are equivalent, so I don't see why we would need to have arbitrary restrictions on how they are persisted.
If the uniqueness constraint is not supported by any relational databases we still have https://github.com/dotnet/efcore/issues/30743

**Comment by roji** [comment]
At the very least I'd prefer HashSet mapping to be split to a separate issue, to gather user feedback on it. Where mapping a .NET Dictionary is very natural (and therefore supported by STJ), mapping a HashSet seems like mostly a theoretical idea to me (i.e. this issue is too abstract). I'd expect lots of people to want Dictionary mapping, and almost nobody to want HashSet mapping.

**Comment by AndriySvyryd** [comment]
IDictionary navigation support is already split into a different issue - https://github.com/dotnet/efcore/issues/21262
Complex IDictionary support would be implemented through a completely different mechanism, so it's also split out - https://github.com/dotnet/efcore/issues/37219

**Comment by roji** [comment]
Thanks. Maybe worth updating the titles to make it clearer what each issue is tracking - they all seem quite similar.
